src/overlays/actors/ovl_Door_Warp1/z_door_warp1.c

```diff
@@ -22,77 +45,1009 @@ const ActorInit Door_Warp1_InitVars = {
+                         Animation_GetLastFrame(&gWarpCrystalAnim), ANIMMODE_ONCE, 0.0f, 1);
+
+    this->skelAnime.curFrame = Animation_GetLastFrame(&gWarpCrystalAnim);
+    this->unk_1AE = 120;
+    this->unk_1B0 = 230;
+    this->warpTimer = 200;
+    this->unk_1B8 = 4000;
+    this->actor.scale.x = this->actor.scale.y = this->actor.scale.z = 1.0f;
+    this->unk_194 = 0.3f;
+    this->unk_198 = 0.3f;
+    this->lightRayAlpha = 0.0f;
+    this->warpAlpha = 0.0f;
+    this->crystalAlpha = 0.0f;
+    this->unk_19C = 0.0f;
+    this->unk_1BC = 1.f;
+    this->actor.shape.yOffset = 800.0f;
+
+    if (gSaveContext.entranceIndex != 0x53) {
+        this->actor.scale.x = 0.0499f;
+        this->actor.scale.y = 0.077f;
+        this->actor.scale.z = 0.09f;
+        this->crystalAlpha = 255.0f;
+    } else {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EV_SHUT_BY_CRYSTAL);
+    }
+    DoorWarp1_SetupAction(this, DoorWarp1_PurpleCrystal);
+}
+
+void DoorWarp1_SetPlayerPos(DoorWarp1* this, GlobalContext* globalCtx) {
+    Player* player = GET_PLAYER(globalCtx);
+
+    player->actor.velocity.y = 0.0f;
+    player->actor.world.pos.x = this->actor.world.pos.x;
+    player->actor.world.pos.y = this->actor.world.pos.y + 55.0f;
+    player->actor.world.pos.z = this->actor.world.pos.z;
+}
+
+void DoorWarp1_BlueCrystal(DoorWarp1* this, GlobalContext* globalCtx) {
+    if (this->warpTimer != 0) {
+        this->warpTimer--;
+    } else {
+        DoorWarp1_SetupAction(this, func_80999214);
+    }
+    DoorWarp1_SetPlayerPos(this, globalCtx);
+}
+
+void func_80999214(DoorWarp1* this, GlobalContext* globalCtx) {
+    s32 temp_f4;
+    f32 phi_f0;
+    s16 phi_v1;
+
+    Math_SmoothStepToF(&this->crystalAlpha, 255.0f, 0.2f, 5.0f, 0.1f);
+
+    phi_f0 = (f32)(40 - this->warpTimer) / 40.0f;
+    phi_f0 = CLAMP_MIN(phi_f0, 0);
+
+    for (phi_v1 = 0; phi_v1 < 3; phi_v1++) {
+        globalCtx->envCtx.unk_8C[0][phi_v1] = globalCtx->envCtx.unk_8C[2][phi_v1] =
+            globalCtx->envCtx.unk_8C[1][phi_v1] = -255.0f * phi_f0;
+    }
+    globalCtx->envCtx.unk_9E = -500.0f * phi_f0;
+
+    this->warpTimer++;
+    if (phi_f0 <= 0) {
+        DoorWarp1_SetupAction(this, func_80999348);
+    }
+    this->actor.shape.rot.y += 0x320;
+    DoorWarp1_SetPlayerPos(this, globalCtx);
+}
+
+void func_80999348(DoorWarp1* this, GlobalContext* globalCtx) {
+    Player* player = GET_PLAYER(globalCtx);
+
+    DoorWarp1_SetPlayerPos(this, globalCtx);
+
+    if (this->warpTimer == 0) {
+        Math_SmoothStepToF(&this->crystalAlpha, 0.0f, 0.1f, 4.0f, 1.0f);
+        if (this->crystalAlpha <= 150.0f) {
+            player->actor.gravity = -0.1f;
+        }
+        if (this->crystalAlpha <= 0.0f) {
+            DoorWarp1_SetupAction(this, func_80999410);
+        }
+    } else {
+        this->warpTimer--;
+    }
+    this->actor.shape.rot.y += 0x320;
+}
+
+void func_80999410(DoorWarp1* this, GlobalContext* globalCtx) {
```

Comment:
`SetPlayerGravity`? Or something similar?

Reply:
This is an action function, so I think ideally it would sound more like an action than as a setter function. I'm thinking `FloatPlayer` might be good?

Reply:
Works for me